python/cuml/cuml/tests/test_incremental_pca.py

```diff
@@ -105,62 +113,69 @@ def test_fit(
+    with using_device_type(device):
+        X, _ = make_blobs(n_samples=nrows, n_features=ncols, random_state=10)
 
-    cu_ipca = cuIPCA(n_components=n_components, whiten=whiten)
+        cu_ipca = cuIPCA(n_components=n_components, whiten=whiten)
 
-    sample_size = int(nrows / batch_size_divider)
-    for i in range(0, nrows, sample_size):
-        cu_ipca.partial_fit(X[i : i + sample_size].copy())
+        sample_size = int(nrows / batch_size_divider)
+        for i in range(0, nrows, sample_size):
+            cu_ipca.partial_fit(X[i : i + sample_size].copy())
 
-    cu_t = cu_ipca.transform(X)
-    cu_inv = cu_ipca.inverse_transform(cu_t)
+        cu_t = cu_ipca.transform(X)
+        cu_inv = cu_ipca.inverse_transform(cu_t)
 
-    sk_ipca = skIPCA(n_components=n_components, whiten=whiten)
+        sk_ipca = skIPCA(n_components=n_components, whiten=whiten)
 
-    X = cp.asnumpy(X)
+        if device == "gpu":
+            X = cp.asnumpy(X)
 
-    for i in range(0, nrows, sample_size):
-        sk_ipca.partial_fit(X[i : i + sample_size].copy())
+        for i in range(0, nrows, sample_size):
+            sk_ipca.partial_fit(X[i : i + sample_size].copy())
 
-    sk_t = sk_ipca.transform(X)
-    sk_inv = sk_ipca.inverse_transform(sk_t)
+        sk_t = sk_ipca.transform(X)
+        sk_inv = sk_ipca.inverse_transform(sk_t)
 
-    assert array_equal(cu_inv, sk_inv, 6e-5, with_sign=True)
+        assert array_equal(cu_inv, sk_inv, 6e-5, with_sign=True)
 
 
-def test_exceptions():
-    X = cupyx.scipy.sparse.eye(10)
-    ipca = cuIPCA()
-    with pytest.raises(TypeError):
-        ipca.partial_fit(X)
+@pytest.mark.parametrize("device", ["gpu", "cpu"])
+def test_exceptions(device):
+    with using_device_type(device):
+        X = global_settings.xsparse.eye(10)
+        ipca = cuIPCA()
+        with pytest.raises(TypeError):
+            ipca.partial_fit(X)
 
-    X = X.toarray()
-    with pytest.raises(NotFittedError):
-        ipca.transform(X)
+        X = X.toarray()
+        with pytest.raises(ValueError):
+            ipca.transform(X)
 
-    with pytest.raises(NotFittedError):
-        ipca.inverse_transform(X)
+        with pytest.raises(ValueError):
+            ipca.inverse_transform(X)
 
-    with pytest.raises(ValueError):
-        cuIPCA(n_components=8).fit(X[:5])
+        with pytest.raises(ValueError):
+            cuIPCA(n_components=8).fit(X[:5])
 
-    with pytest.raises(ValueError):
-        cuIPCA(n_components=8).fit(X[:, :5])
+        with pytest.raises(ValueError):
+            cuIPCA(n_components=8).fit(X[:, :5])
 
 
-def test_svd_flip():
-    x = cp.array(range(-10, 80)).reshape((9, 10))
-    u, s, v = cp.linalg.svd(x, full_matrices=False)
-    u_true, v_true = _svd_flip(u, v, u_based_decision=True)
-    reco_true = cp.dot(u_true * s, v_true)
-    u_false, v_false = _svd_flip(u, v, u_based_decision=False)
-    reco_false = cp.dot(u_false * s, v_false)
+@pytest.mark.parametrize("device", ["gpu", "cpu"])
```

Comment:
We could make a [fixture](https://docs.pytest.org/en/latest/how-to/fixtures.html#how-to-use-fixtures) that we can use in all tests that does this. That way we only have one place to change if we want to add a new device or select a subset of the devices, etc